Add Grafana visualizations folder and histograms doc (#1081)
timescale/docs: timescaledb/tutorials/grafana/visualizations/histograms.md

You can use Grafana to convert 3 distinct formats of data delivery into
histograms. Each comes with its own benefits and challenges:

* Raw data: This method does not require pre-bucketing or aggregating of data.
  Instead, it returns all raw data points in a certain window. This increases
  the accuracy of the histogram, but requires more CPU, memory, and network
  usage, because all bucketing is done in the browser. This could lead to severe
  performance issues.
* Pre-bucketed data: This format contains data that has been pre-bucketed by the
  data source. The Grafana documentation refers to features built into Elastic
  Search's histogram bucket aggregation, or Prometheus' histogram metric as
  examples. The Grafana documentation states that any data source can output
  pre-bucketed data as long as it meets the data format requirements.
* Aggregated data: Grafana also accepts aggregated time bucket data. Such data
  can be easily queried using Timescale's `time_bucket()` hyperfunction. However, this also works with the `date_trunc` postgres function.
  Grafana will automatically select a bucket size (~10% of the total range of

>   data) and bucket the aggregated data appropriately.

Author: @Loquacity got rid of these, I figured it was for stylistic reasons. I'll add them back

```sql
WITH buckets AS (
    SELECT time_bucket('$bucket_interval', time) AS time,
        symbol,
        MAX(day_volume) AS dv_max
    FROM stocks_real_time
    WHERE time >= $__timeFrom()::timestamptz AND time < $__timeTo()::timestamptz
        AND day_volume IS NOT NULL
        AND symbol in ($symbol)
    GROUP BY time_bucket('$bucket_interval', time), symbol
)
SELECT TIME,
    symbol,
    CASE WHEN lag(dv_max ,1) OVER (PARTITION BY symbol ORDER BY time) IS    NULL THEN dv_max
    WHEN (dv_max - lag(dv_max, 1) OVER (PARTITION BY symbol ORDER BY time)) < 0 THEN dv_max
    ELSE (dv_max - lag(dv_max, 1) OVER (PARTITION BY symbol ORDER BY time)) END vol
FROM buckets;
```
This query results in the following histogram:
<img class="main-content__illustration" src="https://assets.timescale.com/docs/images/tutorials/visualizations/histograms/volume_distribution.png" alt="volume distribution"/>
As you can see, we get a left-skewed distribution for the AMZN symbol. In many cases there will
be outliers that represent a few very large volume transactions that distort the distribution. One
solution to this problem is to only look at transactions with a volume less than a certain threshold.

> But that is outside the scope of this example.

Reviewer: For now, let's either cut this sentence (just end it above) OR:

- Show an updated query with a filter based on `vol` (so, probably a having clause) - I don't think this means more screenshots. Just helpful info.
- Mention that one solution would be to use a logarithmic scale but unfortunately this chart doesn't yet support that.

Author: I used the latter in my latest commit ([2dade79](https://github.com/timescale/docs/pull/1081/commits/2dade79f64b405ee543babff11f2c816efebc096)).
Unless you meant for me to use both, then I can add another query with the filter 😄!

  usage, because all bucketing is done in the browser. This could lead to severe
  performance issues.
* **Pre-bucketed data**: This format contains data that has been pre-bucketed by the
  data source. The Grafana documentation refers to features built into Elastic
  Search's histogram bucket aggregation, or Prometheus' histogram metric as
  examples. The Grafana documentation states that any data source can output
  pre-bucketed data as long as it meets the data format requirements.
* **Aggregated data**: Grafana also accepts aggregated time bucket data. Such data
  can be easily queried using Timescale's `time_bucket()` hyperfunction. 
  However, this also works with the `date_trunc` postgres function. Grafana 
  will automatically select a bucket size (~10% of the total range of data) and 
  bucket the aggregated data appropriately.

<highlight type="note">
Histograms are great for analyzing the spread or distribution of data, but they
don't show the change of data over time. If you need to see the distribution of your data over time, try a
[heatmap](https://grafana.com/docs/grafana/latest/visualizations/heatmap/)
instead.
</highlight>

> 

Reviewer 2: ```suggestion

## What you'll learn
This tutorial shows you how to:
*   [Create a price/transaction histogram from raw data](#create-a-price-transaction-histogram-with-raw-data)
*   [Create a price/transaction histogram from pre-aggregated data](#create-a-price-transaction-histogram-with-aggregated-data)
*   [Create a panel showing multiple histograms](#create-a-panel-with-multiple-price-transaction-histograms)
*   [Create a price/volume histogram](#create-a-price-volume-histogram)
```

        AVG(price) avg_price
    FROM stocks_real_time srt
    WHERE symbol = '$symbol'
        AND time >= $__timeFrom()::timestamptz and time < $__timeTo()::timestamptz
    GROUP BY time_bucket('$date_range',time);
    ```

3.  This query yields the following histogram:
    
    <img class="main-content__illustration" src="https://assets.timescale.com/docs/images/tutorials/visualizations/histograms/time_bucket_histogram.png" alt="time bucket histogram"/>

    The second histogram looks similar in shape to the example that uses raw 
    transaction data, but the query returns only around ~1k rows of data with
    each request. This reduces the amount of data over the network and
    processing time in Grafana.

</procedure>

## Price/Transactions Multiple histograms in one panel
In this example we want to compare the distributions of two or more different stocks.
To get the desired effect we need to change our symbol variable from a text variable to a query variable with the multi-value option enabled.

<procedure>

1.  We use this query to fetch all companies in our dataset:

    ```sql
    SELECT DISTINCT symbol FROM company ORDER BY asc;
    ```

> 2.  Create a new dashboard variable with the previous query: 
> 
>     <img class="main-content__illustration" src="https://assets.timescale.com/docs/images/tutorials/visualizations/histograms/symbol_variable_query.png" alt="symbol variable query"/>

Reviewer 2: ```suggestion
1.  Create a new dashboard variable with the previous query: 

        <img class="main-content__illustration" src="https://assets.timescale.com/docs/images/tutorials/visualizations/histograms/symbol_variable_query.png" alt="Screenshot of Grafana showing the Variables > Edit dialog. The variable name is 'symbol', and its type is 'Query'. The Query input field has been filled out with the query from Step 1."/>
```

    If we select more than one value, TimescaleDB returns the transactions for 
    all selected values. Grafana buckets those in separate histograms.
  
3.  Update our main query to use the following query:  

    ```sql
    SELECT time_bucket('$bucket_interval',time) AS time,
        symbol,
        AVG(price) AS avg_price
    FROM stocks_real_time srt
    WHERE symbol in ($symbol)
        AND time >= $__timeFrom()::timestamptz AND time < $__timeTo()::timestamptz
    GROUP BY time_bucket('$bucket_interval',time), symbol;
    ```

4.  When we visualize this query, we get the following histogram:

    <img class="main-content__illustration" src="https://assets.timescale.com/docs/images/tutorials/visualizations/histograms/multiple_histograms.png" alt="multiple histograms"/>

    We can clearly see the 3 distinct histograms but it's impossible to tell 
    them apart from each other. 
    
5.  Click on the green line left of the legend and pick a color:

>     <img class="main-content__illustration" src="https://assets.timescale.com/docs/images/tutorials/visualizations/histograms/color_options.png" alt="color options"/>

Reviewer 2: ```suggestion
    <img class="main-content__illustration" src="https://assets.timescale.com/docs/images/tutorials/visualizations/histograms/color_options.png" alt="Screenshot of the Grafana color picker."/>
```

6.  Doing so results in the following panel:
    
    <img class="main-content__illustration" src="https://assets.timescale.com/docs/images/tutorials/visualizations/histograms/colored_in_histogram.png" alt="colored in histogram"/>

    Here we can clearly see the 3 different price distributions for each stock.

</procedure>

> ## Price/Volume Histogram

Reviewer 2: ```suggestion
## Create a price/volume histogram
```

> An interesting metric to investigate is the trade volume distribution. This 
> distribution gives us insight into the volume at which people buy a stock. 

Reviewer 2: ```suggestion
Besides transaction price, you can also look at trade volumes. The distribution of
trade volume shows you how often and how much people are buying a stock.

```

> The dataset has a column with the daily cumulative volume which we can use to 
> determine the individual transaction volume. To calculate the volume of data for
> each bucket, we need to find the maximum `day_volume` value for the symbol and 
> then subtract each value from the previous bucket maximum value. The difference 
> of these two rows produces the volume within the bucket of time. We can 
> accomplish this with a slightly more complex query that pre-aggregates our data 
> using the TimescaleDB `time_bucket()` hyperfunction and the standard PostgreSQL 
> `MAX()` function, and then use the PostgresSQL `LAG()` window function to 
> subtract each row from the previous, ordered by `time` descending.

Reviewer 2: ```suggestion
The `stocks_real_time` hypertable contains a column with the daily cumulative
volume. You can use this to calculate the volume of data for each bucket. First,
find the maximum `day_volume` value for a symbol within a bucket. Then subtract
each maximum from the previous bucket's maximum. The difference equals the
volume for that bucket.

You can do this with a pre-aggregation query, using:
*   TimescaleDB's `time_bucket` function.
*   PostgreSQL's `max` function.
*   PostgreSQL's `lag` function. Use this to subtract each from from the previous,
    when the rows are ordered by descending `time`.
```
Is descending correct here?

Also suggest linking to API references for these functions
